wled00/colors.cpp

```diff
@@ -70,27 +69,56 @@ uint32_t color_add(uint32_t c1, uint32_t c2, bool fast)
 
 uint32_t color_fade(uint32_t c1, uint8_t amount, bool video)
 {
-  if (c1 == BLACK || amount + video == 0) return BLACK;
+  if (c1 == BLACK || amount == 0) return BLACK;
+  else if (amount == 255) return c1;  
   uint32_t scaledcolor; // color order is: W R G B from MSB to LSB
-  uint32_t r = R(c1);
-  uint32_t g = G(c1);
-  uint32_t b = B(c1);
-  uint32_t w = W(c1);
   uint32_t scale = amount; // 32bit for faster calculation
-  if (video) {
-    scaledcolor  = (((r * scale) >> 8) + ((r && scale) ? 1 : 0)) << 16;
-    scaledcolor |= (((g * scale) >> 8) + ((g && scale) ? 1 : 0)) << 8;
-    scaledcolor |=  ((b * scale) >> 8) + ((b && scale) ? 1 : 0);
-    scaledcolor |= (((w * scale) >> 8) + ((w && scale) ? 1 : 0)) << 24;
-  } else {
-    scaledcolor  = ((r * scale) >> 8) << 16;
-    scaledcolor |= ((g * scale) >> 8) << 8;
-    scaledcolor |=  (b * scale) >> 8;
-    scaledcolor |= ((w * scale) >> 8) << 24;
+  uint32_t addRemains = 0;
+  if (!video) amount++; // add one for correct scaling using bitshifts
+  else { // video scaling: make sure colors do not dim to zero if they started non-zero
+    addRemains = R(c1) ? 0x00010000 : 0;
+    addRemains |= G(c1) ? 0x00000100 : 0;
+    addRemains |= B(c1) ? 0x00000001 : 0;
+    addRemains |= W(c1) ? 0x01000000 : 0;
   }
+  uint32_t rb = (((c1 & 0x00FF00FF) * scale) >> 8) & 0x00FF00FF; // scale red and blue
+  uint32_t wg = (((c1 & 0xFF00FF00) >> 8) * scale) & 0xFF00FF00; // scale white and green
+  scaledcolor = (rb | wg) + addRemains;  
   return scaledcolor;
 }
 
+// 1:1 replacement of fastled function optimized for ESP, slightly faster, more accurate and uses less flash (~ -200bytes)
```

Comment:
If you do that, why not integrate this into WLED native `Segment::color_from_palette()`? Just a thought.
The downside: Some effects would need to be rewritten.

And BTW I do not like how FastLED wraps palettes. IMO they should never be wrapped. If I want "round" palette I should create one that starts and stops with the same color.

Reply:
that is a valid point. I just wanted to see if I can make the function faster as a test and found that a 32bit native version is not only faster, it also is more compact and more accurate (fastled uses dirty tricks).
I think wrapping is a valid feature. many of the palettes are not "round". The palette FX is a good example where it is useful: it give a "hard color cut" if palettes are not wrapped.

Reply:
If it's ok with you, I would leave it as-is for now. Integrating it into `Segment::color_from_palette()` would require a lot of code changes all over, even in usermods.

Reply:
Yes, we can do that.